[ConstraintSystem] Relax the left-over information check in `ComponentStep`

xedin opened the pull request:
Instead of crashing in release builds, let's simply fail the
step and let type-checker produce a fallback diagnostic.

Resolves: rdar://56167233

<!--
Before merging this pull request, you must run the Swift continuous integration tests.
For information about triggering CI builds via @swift-ci, see:
https://github.com/apple/swift/blob/master/docs/ContinuousIntegration.md#swift-ci

Thank you for your contribution to Swift!
-->

xedin committed:
[ConstraintSystem] Relax the left-over information check in `ComponentStep`

Instead of crashing in release builds, let's simplify fail the
step and let type-checker produce a fallback diagnostic. Maintain
a trap in debug builds to make it easier to investigate the root
cause of the issue.

Resolves: rdar://56167233

hamishknight commented on lib/Sema/CSStep.cpp:
```diff
@@ -362,22 +362,60 @@ StepResult ComponentStep::take(bool prevFailed) {
+#endif
     }
+
     if (CS.hasFreeTypeVariables()) {
+#ifndef NDEBUG
       CS.print(llvm::errs());
       llvm::report_fatal_error("Free type variables left over?");
+#else
+      if (CS.isDebugMode()) {
+        getDebugLogger() << "(failed due to remaining free type variables)\n";
+        CS.print(getDebugLogger(/*indent=*/false));
+      }
+
+      return finalize(/*isSuccess=*/false);
+#endif
     }
```
I think this entire condition is actually dead code, as it's already handled in an above `if` statement, I guess I must have missed that when adding it

xedin replied:
Ah, I see yeah, let's remove it.

xedin replied:
Fixed!

xedin committed:
[ConstraintSystem] Remove check for left-over free type variables since it's handle gracefully

xedin commented:
@hamishknight I have removed redundant check you mentioned and added `InvalidState` flag to make sure that solving is aborted early and solutions are never produced, this should make sure that solving in release mode always ends up with a fallback diagnostic.

hamishknight reviewed (approved):
Thanks!

hborla reviewed (approved):
Good idea to produce an error message instead 👍

hborla commented:
@swift-ci please smoke test

xedin commented:
@hamishknight @hborla After thinking about this some more I think we can drop abort blocks all together because there it would be impossible now to miscompile, WDYT?

xedin committed:
[ConstraintSystem] Add `InvalidState` flag to indicate that any attempt to solve should fail

As soon as `InvalidState` flag is set solving of the constraint
system as aborted and all subsequent calls to `solveImpl` would
produce no solutions.

hborla commented:
@xedin Yeah, I think it's fine to always report the error message instead of aborting. If we still want it to crash in this part of the code in a debug build, we could just put in an assert

xedin committed:
[CSStep] Don't abort upon discovering constraint system in an invalid state

Setting `InvalidState` already makes sure that there is no possibility
of miscompile.

xedin commented:
@swift-ci please smoke test

xedin commented:
Adjusted that, for debugging it would print the constraints in question so I just removed NDEBUG blocks.

hamishknight commented:
I'd prefer if we crashed in a debug build to make it loud and obvious that we have a logic error, but yeah that could be done with an assert.

xedin commented:
I think asking to file a bug is loud enough, that would also indicate where the error is better than crash report because not all tools capture stacktraces.

hamishknight commented:
Hmm, won't we emit "type of expression is ambiguous without more context" for an expression with no solutions?

xedin committed:
[ConstraintSystem] Make sure that invalid state is always diagnosed via fallback diagnostic

xedin commented:
Ugh, you are right! I have completely forgot about that message assuming that `forUndiagnosedError` would always result in `failed to produce a diagnostic ...` message. Fixed now.

xedin commented:
@swift-ci please smoke test

xedin committed:
[ConstraintSystem] Don't attempt to rollback state and scope from incorrect state

xedin commented:
@swift-ci please smoke test

xedin commented:
@swift-ci please smoke test
